Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp

```diff
@@ -1797,14 +1797,33 @@ void MediaPlayerPrivateGStreamer::handleMessage(GstMessage* message)
         asyncStateChangeDone();
         break;
     case GST_MESSAGE_STATE_CHANGED: {
-        if (!messageSourceIsPlaybin || m_isDelayingLoad)
-            break;
-
         GstState newState;
         gst_message_parse_state_changed(message, &currentState, &newState, nullptr);
-
         GST_DEBUG_OBJECT(pipeline(), "Changed state from %s to %s", gst_element_state_get_name(currentState), gst_element_state_get_name(newState));
 
+#if USE(GSTREAMER_HOLEPUNCH)
+        if (currentState == GST_STATE_NULL && newState == GST_STATE_READY) {
+            // If we didn't create a video sink, store a reference to the created one.
+            if (!m_videoSink) {
+                // Detect the videoSink element. Getting the video-sink property of the pipeline requires
+                // locking some elements, which may lead to deadlocks during playback. Instead, identify
+                // the videoSink based on its metadata.
+                GstElement* element = GST_ELEMENT(GST_MESSAGE_SRC(message));
+                if (GST_IS_BASE_SINK(element)) {
+                    const gchar* klassStr = gst_element_get_metadata(element, "klass");
+                    if (strstr(klassStr, "Sink") && strstr(klassStr, "Video")) {
+                        m_videoSink = element;
+
+                        // Ensure that there's a buffer with the transparent rectangle available when playback is going to start.
+                        pushNextHolePunchBuffer();
+                    }
+                }
+            }
```

Comment:
Isn't this... dead code? In which case is `m_videoSink` null, exactly? I see `createVideoSink()` unconditionally called from `createGSTPlayBin()`.

Reply:
Ok I see now this is called only for `USE(WPEWEBKIT_PLATFORM_BCM_NEXUS)`. Can you add it to the ifdef?